src/exodus/test/wallet_tests.cpp

```diff
@@ -0,0 +1,199 @@
+    BOOST_CHECK(!retrieved.isUsed);
+}
+
+BOOST_AUTO_TEST_CASE(update_and_delete_entry_from_chain)
+{
+    auto entry = createAndGetEntry(1, 1);
+    BOOST_CHECK(wallet->UpdateSigmaMint(entry.GetId(), 10, 1600, 100));
+    auto beforeUpdated = wallet->GetSigmaEntry(entry.GetId());
+    BOOST_CHECK_EQUAL(10, beforeUpdated.groupId);
+    BOOST_CHECK_EQUAL(1600, beforeUpdated.index);
+    BOOST_CHECK_EQUAL(100, beforeUpdated.block);
+
+    BOOST_CHECK(wallet->ClearSigmaMintChainState(entry.GetId()));
+
+    auto updated = wallet->GetSigmaEntry(entry.GetId());
+    BOOST_CHECK_EQUAL(0, updated.groupId);
+    BOOST_CHECK_EQUAL(0, updated.index);
+    BOOST_CHECK_EQUAL(-1, updated.block);
+    BOOST_CHECK(!updated.isUsed);
+}
+
+BOOST_AUTO_TEST_CASE(make_entry_as_used)
+{
+    auto entry = createAndGetEntry(1, 1);
+    BOOST_CHECK(wallet->SetSigmaMintUsedStatus(entry.GetId(), true));
+    auto updated = wallet->GetSigmaEntry(entry.GetId());
+    BOOST_CHECK(updated.isUsed);
+
+    BOOST_CHECK(wallet->SetSigmaMintUsedStatus(entry.GetId(), false));
+    updated = wallet->GetSigmaEntry(entry.GetId());
+    BOOST_CHECK(!updated.isUsed);
+}
+
+BOOST_AUTO_TEST_CASE(list_entry_no_coins)
+{
+    std::list<exodus::SigmaEntry> listSigma;
+    wallet->ListSigmaEntries(listSigma);
+    BOOST_CHECK_EQUAL(0, listSigma.size());
+}
+
+BOOST_AUTO_TEST_CASE(list_entry_have_two_coins_same_property)
+{
+    auto pubKey = wallet->CreateSigmaMint(1, 1);
+    auto pubKey2 = wallet->CreateSigmaMint(1, 1);
+
+    std::list<exodus::SigmaEntry> listSigma;
+    wallet->ListSigmaEntries(listSigma);
+    BOOST_CHECK_EQUAL(2, listSigma.size());
+
+    auto commitment = pubKey.GetCommitment();
+    auto commitment2 = pubKey2.GetCommitment();
+
+    auto first = listSigma.front().getPublicKey().GetCommitment();
+    listSigma.pop_front();
+    auto second = listSigma.front().getPublicKey().GetCommitment();
+    BOOST_CHECK(
+        (commitment == first && commitment2 == second)
+        || (commitment == second && commitment2 == first)
+    );
+}
+
+BOOST_AUTO_TEST_CASE(list_entry_have_two_coins_different_property)
+{
+    auto prop1Entry = createAndGetEntry(1, 1);
+    auto prop2Entry = createAndGetEntry(2, 1);
+
+    std::list<exodus::SigmaEntry> listSigma;
+    wallet->ListSigmaEntries(listSigma);
+    BOOST_CHECK_EQUAL(2, listSigma.size());
+
+    std::list<exodus::SigmaEntry> listProp1Sigma;
+    wallet->ListSigmaEntries(1, listProp1Sigma);
+    BOOST_CHECK_EQUAL(1, listProp1Sigma.size());
+    BOOST_CHECK(prop1Entry == listProp1Sigma.front());
+
+    std::list<exodus::SigmaEntry> listProp2Sigma;
+    wallet->ListSigmaEntries(2, listProp2Sigma);
+    BOOST_CHECK_EQUAL(1, listProp2Sigma.size());
+    BOOST_CHECK(prop2Entry == listProp2Sigma.front());
+}
+
+BOOST_AUTO_TEST_CASE(is_exists)
+{
+    auto existsEntry = createAndGetEntry(1, 1);
+
+    exodus::SigmaPrivateKey nonExistsPriv;
+    do {
+        nonExistsPriv.Generate();
+    } while (!nonExistsPriv.IsValid());
```

Comment:
As my previous comment that we don't need this loop if we change `SigmaPrivateKey` to never generate invalid key.